From 8ccddd70aa1f94a714ac1d95f419022728545f4a Mon Sep 17 00:00:00 2001
From: ginugeorge <ginugeorge@ami.com>
Date: Fri, 8 Sep 2023 18:57:05 +0530
Subject: [PATCH 81/81] Closing SSE stream when Subscription is deleted

When Subscription is deleted by HTTP Delete request to the
subscription URI, clossing the assosciated SSE stream.

Tested: Verified that the SSE stream gets closed when the subscription is
deleted. All other event service functionality is working as earlier.

Signed-off-by: ginugeorge <arunthomasb@ami.com>
---
 redfish-core/include/event_service_manager.hpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/redfish-core/include/event_service_manager.hpp b/redfish-core/include/event_service_manager.hpp
index b4b8934b..dede2017 100644
--- a/redfish-core/include/event_service_manager.hpp
+++ b/redfish-core/include/event_service_manager.hpp
@@ -428,6 +428,12 @@ class Subscription : public persistent_data::UserSubscription
 
     ~Subscription() = default;
 
+    void getSseConnection(std::shared_ptr<crow::SseConnection>& connPtr)
+    {
+	connPtr = sseConn;
+	return;
+    }	
+
     bool sendSNMPTrap(uint32_t eventId, uint64_t timestamp, uint8_t sev, std::string &msg)
     {
 	persistent_data::EventServiceConfig eventServiceConfig =
@@ -1148,8 +1154,15 @@ class EventServiceManager
     void deleteSubscription(const std::string& id)
     {
         auto obj = subscriptionsMap.find(id);
+	std::shared_ptr<crow::SseConnection> sseConnPtr = NULL;
         if (obj != subscriptionsMap.end())
         {
+	    std::shared_ptr<Subscription> entry = obj->second;
+	    if (entry->subscriptionType == subscriptionTypeSSE)
+	    {
+		entry->getSseConnection(sseConnPtr);
+	    }
+
             subscriptionsMap.erase(obj);
             auto obj2 = persistent_data::EventServiceStore::getInstance()
                             .subscriptionsConfigMap.find(id);
@@ -1168,6 +1181,10 @@ class EventServiceManager
                                 "OpenBMC.0.1.EventSubscriptionRemoved",
                                 "REDFISH_MESSAGE_ARGS=%s", id.c_str(), NULL);
             }
+	    if(sseConnPtr)
+	    {
+		sseConnPtr->close("subscription deleted");
+	    }
         }
     }
 
-- 
2.39.2

